core-tests/shared/src/test/scala/zio/stm/ZSTMSpec.scala

```diff
@@ -1392,6 +1392,33 @@ object ZSTMSpec extends ZIOBaseSpec {
 
         assertM(tx.commit)(equalTo((Left("error"), "error")))
       } @@ zioTag(errors)
+    ),
+    suite("issue #4786")(
+      testM("check atomicity ") {
+        val max = 99
+        def transaction(i: Int, ref: TRef[Int], sum: TRef[Int]) =
+          for {
+            ref_prev <- ref.get
+            sum_prev <- sum.get
+            _        <- sum.set(sum_prev + ref_prev + i)
+            _        <- ref.set(i)
+          } yield i
+
+        for {
+          (sum, ref) <- ZSTM.atomically(TRef.make[Int](0) zip TRef.make[Int](0))
+
+          fibers <- ZIO.forkAll {
+                      scala.util.Random.shuffle(1 to max).map { i =>
+                        ZSTM.atomically(transaction(i, ref, sum)).fork
+                      }
+                    }
```

Comment:
I don't think this is necessary an issue with STM. This part involves nested forking, which can be treated as an antipattern. Dropping it or replacing with forkDaemon would cause the test to pass.

cc @adamgfraser

Reply:
@FrancisToth There is an issue with your test.

Notice that you are actually forking two levels of fibers here. `forkAll` forks one set of fibers and then each of those does as its only working forking another set of fibers which are supposed to perform the actual STM transaction.

This creates a number of issues unrelated to STM. First, due to fiber supervision when the child fiber completes execution it attempts to interrupt its children, so the grandchild fibers that are actually supposed to do the STM transaction may be interrupted during or even before they begin execution, since the child fibers do no work other than forking the child.

Second, this means that when the fibers are joined on the line below we do not have any guarantee that the work has actually been completed.

The easiest way to fix this test is to remove the extraneous `fork` entirely like this:

```scala
          fibers <- ZIO.forkAll {
                      scala.util.Random.shuffle(1 to max).map { i =>
                        ZSTM.atomically(transaction(i, ref, sum))
                      }
                    }
          _ <- fibers.join
```

In this case the test passes.

If we do want to fork the second level of fibers we would have to fork them as daemon fibers to make sure they are not interrupted due to auto supervision and we would also have to join back both the children and grandchildren to ensure that they have completed execution like this:

```scala
           fibers <- ZIO.forkAll {
                       scala.util.Random.shuffle(1 to max).map { i =>
                         ZSTM.atomically(transaction(i, ref, sum)).forkDaemon
                       }
                     }
           _ <- fibers.join.flatMap(fibers => ZIO.foreach(fibers)(_.join))
```

In this case the test passes as well.

So I think we can close this PR and the associated issue.

Copying @mijicd.

Reply:
Thanks a lot @mijicd and @adamgfraser  for the details and your time.
I'm sorry this is a clear mistake on my part.